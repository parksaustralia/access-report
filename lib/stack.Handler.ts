import { Settings, DateTime } from "luxon";
import axios from "axios";
import Handlebars from "handlebars";
import he from "he";
import sendgrid from "@sendgrid/client";

declare var process: {
  env: {
    SENDGRID_API_KEY: string;
  };
};

sendgrid.setApiKey(process.env.SENDGRID_API_KEY);

Settings.defaultZoneName = "Australia/Darwin";

async function getData() {
  const response = await axios.request({
    method: "GET",
    url: "https://parksaustralia.gov.au/kakadu/access/feed.xml",
  });

  return response.data as string;
}

async function getTemplate() {
  const response = await sendgrid.request({
    method: "GET",
    url: "/v3/designs/3aaf2986-a36d-4a7d-883b-28cb497eb56f",
  });
  return response[1].html_content;
}

async function postSingleSend(htmlContent: String, formattedDate: String) {
  const response = await sendgrid.request({
    method: "POST",
    url: "/v3/marketing/singlesends",
    body: JSON.stringify({
      email_config: {
        editor: "design",
        generate_plain_content: false,
        html_content: htmlContent,
        sender_id: 1119454,
        subject: `Kakadu access report: ${formattedDate}`,
        suppression_group_id: 15891,
      },
      name: `Kakadu access report - ${formattedDate}`,
      send_to: {
        list_ids: ["fc55886e-c778-41d8-a0d8-83cde65deaa2"],
      },
      status: "scheduled",
    }),
  });

  return response;
}

async function scheduleSingleSend(messageId: String, sendAt: String) {
  const response = await sendgrid.request({
    method: "PUT",
    url: `/v3/marketing/singlesends/${messageId}/schedule`,
    body: JSON.stringify({ send_at: sendAt }),
  });

  return response;
}

export async function handler() {
  const date = DateTime.local();
  const formattedDate = date.toFormat("d MMMM yyyy");

  let xml: string;
  try {
    xml = await getData();
  } catch (error) {
    console.log("Failed to download data from parksaustralia.gov.au:");
    console.log(error);
    return false;
  }

  const regex = /<content type="html">([\s\S]*?)<\/content>/g;
  const rawHtml = regex.exec(xml);

  if (rawHtml == null) {
    console.log("Failed to parse XML");
    return false;
  }

  const html = he.decode(rawHtml[1]);

  let template: String;
  try {
    template = await getTemplate();
  } catch (error) {
    console.log("Failed to download template:");
    console.log(error);
    return false;
  }

  let handlebars: HandlebarsTemplateDelegate;
  let htmlContent: String;

  try {
    handlebars = Handlebars.compile(template);
    htmlContent = handlebars({ formattedDate, html });
  } catch (error) {
    console.log("Failed to compile template:");
    console.log(error);
    return false;
  }

  let singleSendResponse;

  try {
    singleSendResponse = await postSingleSend(htmlContent, formattedDate);
  } catch (error) {
    console.log("Failed to post single send to SendGrid:");
    console.log(error);
    return false;
  }

  type singleSendResponseBody = {
    id: String;
  };

  try {
    const messageId = (singleSendResponse[0].body as singleSendResponseBody).id;
    const sendAt = DateTime.local().startOf("day").plus({ hours: 3 }).toISO();
    const response = await scheduleSingleSend(messageId, sendAt);
  } catch (error) {
    console.log("Failed to schedule single send in SendGrid:");
    console.log(error);
    console.log(error?.response?.body?.errors);
    return false;
  }

  console.log("Done");
  return true;
}

handler();
