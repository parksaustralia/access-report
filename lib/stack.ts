import { LambdaFunction } from "@aws-cdk/aws-events-targets";
import { Rule, Schedule } from "@aws-cdk/aws-events";
import * as cdk from "@aws-cdk/core";
import * as lambda from "@aws-cdk/aws-lambda";
import * as lambdaNode from "@aws-cdk/aws-lambda-nodejs";
import * as secretsManager from "@aws-cdk/aws-secretsmanager";

export class Stack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const handler = new lambdaNode.NodejsFunction(this, "Handler", {
      environment: {
        SENDGRID_API_KEY: `${
          secretsManager.Secret.fromSecretNameV2(
            this,
            "sendgridApiKey",
            "accessReport/sendgridApiKey"
          ).secretValue
        }`,
      },
      memorySize: 3008,
      runtime: lambda.Runtime.NODEJS_12_X,
      timeout: cdk.Duration.seconds(30),
      tracing: lambda.Tracing.ACTIVE,
    });

    const target = new LambdaFunction(handler);

    // Execute daily at 12 pm Darwin time
    new Rule(this, "ScheduleRule", {
      schedule: Schedule.cron({ minute: "30", hour: "12" }),
      targets: [target],
    });
  }
}
