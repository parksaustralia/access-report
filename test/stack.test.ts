import { Stack } from "../lib/stack";
import { SynthUtils } from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";

test("matches snapshot", () => {
  const app = new cdk.App();
  const stack = new Stack(app, "MyTestStack");
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
