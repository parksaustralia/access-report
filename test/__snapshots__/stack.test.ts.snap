// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0ArtifactHash656347CC": Object {
      "Description": "Artifact hash for asset \\"0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0\\"",
      "Type": "String",
    },
    "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0S3Bucket4656C5AC": Object {
      "Description": "S3 bucket for asset \\"0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0\\"",
      "Type": "String",
    },
    "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0S3VersionKeyD4F16D4A": Object {
      "Description": "S3 key for asset version \\"0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Handler886CB40B": Object {
      "DependsOn": Array [
        "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "HandlerServiceRoleFCDC14AE",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0S3Bucket4656C5AC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0S3VersionKeyD4F16D4A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0925b907aeaa6f84a7424eb3bb30d1b4fe931f417dea631e6e028a14026bc4f0S3VersionKeyD4F16D4A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SENDGRID_API_KEY": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{{resolve:secretsmanager:arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":secretsmanager:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":secret:accessReport/sendgridApiKey:SecretString:::}}",
                ],
              ],
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 3008,
        "Role": Object {
          "Fn::GetAtt": Array [
            "HandlerServiceRoleFCDC14AE",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "HandlerServiceRoleDefaultPolicyCBD0CC91": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "Roles": Array [
          Object {
            "Ref": "HandlerServiceRoleFCDC14AE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HandlerServiceRoleFCDC14AE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleRuleAllowEventRuleMyTestStackScheduleRule3625033547FCA01B": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "ScheduleRuleDA5BD877",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduleRuleDA5BD877": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(30 12 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "Handler886CB40B",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
